src/main/resources/templates/java/test/test/${packageNameFolder}/testutils/ClassNameScanner.java

```diff
@@ -0,0 +1,235 @@
+									+ " Please check for wrong upper case / lower case lettering."
+							: "The exercise expects a class with the name " + expectedClassName + " in the package " + expectedPackageName
+									+ ". We found that you implemented a class " + observedClassName + ", in the package " + observedClassName
+									+ ", which deviates from the expectation."
+									+ " Please check for wrong upper case / lower case lettering and make sure you place it in the correct package.";
+					break;
+				} else if(hasTypos) {
+					scanResultType = classIsCorrectlyPlaced 
+							? ScanResultType.TYPOS_CORRECTPLACE
+							: ScanResultType.TYPOS_MISPLACED;
+					scanResultMessage = classIsCorrectlyPlaced
+							? "The exercise expects a class with the name " + expectedClassName
+									+ ". We found that you implemented a class " + observedClassName + ", which deviates from the expectation."
+									+ " Please check for typos in the class name."
+							: "The exercise expects a class with the name " + expectedClassName + "in the package " + expectedPackageName
+									+ ". We found that you implemented a class " + observedClassName + ", in the package " + observedPackageName
+									+ ", which deviates from the expectation."
+									+ " Please check for typos in the class name and make sure you place it in the correct package.";
+					break;
+				} else {
+					scanResultType = ScanResultType.NOTFOUND;
+					scanResultMessage = "You have implemented " + observedClassName + " in the package " + observedClassName
+							+ ". This class is not expected in the exercise."
+							+ "\n The assignment expects the following classes and in the according packages: \n";
+					
+					// Append the expected classes and package names to the result message
+					for(String expectedClassName : expectedClasses.keySet()) {
+						scanResultMessage += "Class: " + expectedClassName +
+											" Package:" + expectedClasses.get(expectedClassName) +",\n";
+					} 
+				}
+			}
+		}
+		
+		return new ScanResult(scanResultType, scanResultMessage);
+	}
+	
+	/**
+	 * This method retrieves the actual type names and their packages by walking the project file structure.
+	 * The root node (which is the assignment folder) is defined in the pom.xml file of the project.
+	 * @return The JSON object containing the type names as keys and the type packages as values.
+	 */
+	private HashMap<String, String> retrieveObservedClasses() {
+		HashMap<String, String> observedTypes = new HashMap<String, String>();
+		
+		try {
+			File pomFile = new File("pom.xml");
+			DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+			DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+			Document pomDocument = documentBuilder.parse(pomFile);
+						
+			NodeList buildNodes = pomDocument.getElementsByTagName("build");
+			for(int i = 0; i < buildNodes.getLength(); i++) {
```

Comment:
Can we use a for each loop here? This might be simpler and less error prone. (Depends if NodeList supports this)

Reply:
Yup NodeList does actually support this, changed it.